docs/proposals/20220414-runtime-hooks.md

```diff
@@ -0,0 +1,426 @@
+* [Error management](https://github.com/kubernetes-sigs/cluster-api/blob/b48a6ed07ac2bd353f99000270510369f4baa1a5/docs/proposals/20220221-runtime-SDK.md#error-management)
+* [Avoid dependencies](https://github.com/kubernetes-sigs/cluster-api/blob/b48a6ed07ac2bd353f99000270510369f4baa1a5/docs/proposals/20220221-runtime-SDK.md#avoid-dependencies)
+
+
+###  Security Model
+
+For the general Runtime Extension security model please refer to the [developer guide in the Runtime SDK proposal](https://github.com/kubernetes-sigs/cluster-api/blob/75b39db545ae439f4f6203b5e07496d3b0a6aa75/docs/proposals/20220221-runtime-SDK.md#security-model).
+
+###  Risks and Mitigations
+
+####  Runtime Extension blocks Cluster lifecycle indefinitely
+
+Cluster lifecycle can be blocked indefinitely when a Runtime Extension either blocks or fails indefinitely. Mitigation:
+
+* A Runtime Extension should be extensively unit and e2e tested to ensure it behaves as expected.
+
+
+## Alternatives
+
+Alternatives to Runtime Hooks for comprehensive add-on management in Cluster API include:
+
+### External components watching CAPI resources without hooks
+
+This is the current pattern used by Cluster Resource Set. The implementers can only react to changes on CAPI resources, e.g. a Cluster being created, but they have no control over the cluster lifecycle.
+
+This and similar solutions based on scripting or git ops approaches are considered inadequate for comprehensive, integrated add-on management as they have limited insight into the state of the Cluster and can not easily influence reconciliation based on add-on state.
+
+More details about this in the [Cluster Addon Proposal](https://docs.google.com/document/d/1TdbfXC2_Hhg0mH7-7hXcT1Gg8h6oXKrKbnJqbpFFvjw/edit).
+
+### OpenAPI spec implementation alternatives
+
+For the implementation detail of Open API spec, we considered following alternative approaches:
+#### Adding Cluster info to request vs providing only the Cluster name
+
+In the proposed Open API specification for request type we have a full Cluster object. We considered an alternative of only including the Cluster name and namespace to reduce the size of the message. It was rejected based on the assumption that most extensions will require at least some additional information from the Cluster. Sending the full object reduces calls to the API server.
+
+#### Embedding CAPI types in request vs using runtime.RawExtension
+
+In the proposed Open API specification we are including the Cluster API object in requests. We considered using runtime.RawExtension in order to avoid having to bump the version of lifecycle hooks when bumping the version of the CAPI types. It was rejected as sending another version of the CAPI type via runtime.RawExtension would always be a breaking change. Embedding the type directly makes the version of the API used explicit.
```

Comment:
Thinking a little bit more about this, we should consider how to trigger down-conversion of the cluster object when calling an older version of the hook.
Most probably this requires some thinking/experimenting about implementation details, but I think the current working assumption should not be put in discussion for this

Reply:
Good point about the down-conversion of the Cluster object. This will need more experimenting to identify the correct implementation for this. Shall I drop the `Embedding CAPI types in request vs using runtime.RawExtension` section from the proposal?

Reply:
No, let's stick to the current working assumptions, but we should track this issue to be investigated